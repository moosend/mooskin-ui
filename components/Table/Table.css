@import '../_utils/globals/screens.css';

.table{
    width: 100%;
    border-collapse: collapse;
}

.tbody{
    border: 1px solid #eaeaea;
}

.th{
    background-color: #efefef;
    padding-top: 8px;
    padding-bottom: 8px;
    color: #aaaaaa;
    font-size: .8rem;
    border-bottom: 1px solid #d4d4d4;
    font-family: Arial, Helvetica, sans-serif;
    font-weight: normal;
}

.td{
    text-align: center;
    vertical-align: middle;
    padding-top: 20px;
    padding-bottom: 20px;
    font-size: 1rem;
    border-bottom: 1px solid #eaeaea;
    color:#6b6b6b;
}

.heading {
    display: none;
}

.th{
    cursor: pointer;
}

@media (--deviceXsmall){
    .thead{
        display: none;
    }

    .tr{
        display: flex;
        flex-direction: column;
        border: 3px solid #eaeaea;
        box-shadow: 10px 10px 30px #eaeaea;
    }

    .tr + .tr{
        margin-top: 50px;
    }

    .heading{
        display: block;
        font-weight: 800;
        flex: 1;
        text-align: left;
    }

    .tdContent{
        display: flex;
        flex-direction: row;
        font-size: 3vw;
    }

    .td{
        border: none;
        padding: 20px;
    }

    .td + .td {
        border-top: 1px solid #eaeaea
    }

    /* .td:nth-child(even){
        background: #eaeaea;
    } */

    .content{
        flex: 1;
        text-align: left;
        font-size: 3vw;
    }
    
    .tbody{
        border: none;
    }
}