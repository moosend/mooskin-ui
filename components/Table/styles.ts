import styled from 'styled-components';

// Models
import { IBoxComponentProps } from '../Box/model';

// Components
import { Box } from '../Box/Box';

// "CSS" variables
import variables from '../_utils/globals/variables';

export const StyledTable = styled(Box)<IBoxComponentProps>`
	border: solid 1px #e2e2e2;
	background-color: ${(props) => props.palette?.backgroundColors.white || variables.backgroundColors.white};
	display: flex;
	flex-direction: column;
`;

StyledTable.displayName = 'StyledTable';

export const StyledTableHeader = styled(Box)<IBoxComponentProps>`
	display: flex;
	align-items: center;
	min-height: 45px;
`;

StyledTableHeader.displayName = 'StyledTableHeader';

export const StyledTableHeaderItem = styled(Box)<IBoxComponentProps>`
	font-size: 11px;
	font-weight: bold;
	font-stretch: normal;
	font-style: normal;
	letter-spacing: normal;
	text-align: left;
	color: ${(props) => props.palette?.fontColors.secondary2 || variables.fontColors.secondary2};
	display: flex;
	align-items: center;
`;

StyledTableHeaderItem.displayName = 'StyledTableHeaderItem';

export const StyledTableRow = styled(Box)<IBoxComponentProps>`
	display: flex;
	align-items: center;
	min-height: 90px;
	border-top-width: 1px;
	border-top-style: solid;
	border-top-color: ${(props) => props.palette?.borderColors.gray1 || variables.borderColors.gray1};
	border-bottom-width: 1px;
	border-bottom-style: solid;
	border-bottom-color: ${(props) => props.palette?.borderColors.gray1 || variables.borderColors.gray1};
	:last-child {
		border: none;
	}
`;

StyledTableRow.displayName = 'StyledTableRow';

export const StyledTableRowItem = styled(Box)<IBoxComponentProps>`
	font-size: 16px;
	font-weight: 500;
	font-stretch: normal;
	font-style: normal;
	letter-spacing: normal;
	text-align: left;
	color: ${(props) => props.palette?.fontColors.text || variables.fontColors.text};
	display: flex;
	align-items: center;
`;

StyledTableRowItem.displayName = 'StyledTableRowItem';
