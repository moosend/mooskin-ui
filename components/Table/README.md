# MooSkin UI - Table Component

The MooSkin Button Component works similar to the normal HTML `<table/>` element, but with pre-defined styling and attributes.

___

### Usage

To start using the Table Component first you have to Import it

```
Import {Table, TR, TH, TD, THead, TBody} from 'mooskin';
```
or
```
import Table from 'mooskin/lib/Table';
```

Styles can also be imported this way, if they were not imported globaly and you want to import just the table styles to keep the css bundle size healthy: 

```
import 'mooskin/lib/Table/styles.css';
```

And then you can simply start using it by typing

```
<Table attribute1="atr" attribute2="asd" >
    <THead>
        <TR>
            <TH>header 1</TH>
            <TH>header 2</TH>
            <TH>header 3</TH>
        </TR>
    </THead>
    <TBody>
        <TR>
            <TD>td 1</TD>
            <TD>td 2</TD>
            <TD>td 3</TD>
        </TR>
    </TBody>
</Table>
```
As you can see only the first `<Table>` element is a composite component, as its children are just the usual dom element with capital letters.


For easy use, the components are named similar to normal HTML components but with a capital first letter.

Like the `<table/>` element it will accept given attributes and render differently based on the given attributes

<div class="playground-doc">

## Supported attributes 

* `id` - id of the element
* `className` - css class
* `style` - table styles

Allthough these attributes are supported, all of them are optional. normal `<table>` props can be used as well(like border, cellpadding etc.). They will just be passed to the underlying `<table>`. 

</div>

#### For more

___

[MooSkin-UI](https://github.com/moosend/mooskin-ui)
