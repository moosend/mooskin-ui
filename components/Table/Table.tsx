import * as React from 'react';

import styles from './Table.css';

export interface ITableProps{

    /** override Table id */
    id?: string;

    /** Table class */
    className?: string;

    /** override Table styles */
    style?: React.CSSProperties;

    children?: any;

}

export interface IHeaderProps extends ITableProps {

    onClick?: (e: React.MouseEvent<HTMLElement>) => void;
}

export default class Table extends React.Component<ITableProps, {}> {

    public static defaultProps: Partial<ITableProps> = {
        className: '',
        style: {}
    };

    public static THead: React.StatelessComponent<ITableProps>;
    public static TH: React.StatelessComponent<IHeaderProps>;
    public static TBody: React.StatelessComponent<ITableProps>;
    public static TR: React.StatelessComponent<ITableProps>;
    public static TD: React.StatelessComponent<ITableProps>;

    public render(){

        const {newHeaders, newBody} = this.getContent();

        return (
            <div>
                 <table {...this.props} className={`${styles.table} ${this.props.className}`}>
                    {newHeaders}
                    {newBody}
                </table>
            </div>
        );
    }

    private getContent = () => {

        let headStrings: string[] = [];
        let newHeaders: React.ReactElement<ITableProps> = <TR/>;
        let newCells: Array<React.ReactElement<ITableProps>> = [];
        const newRows: Array<React.ReactElement<ITableProps>> = [];
        const newBody: Array<React.ReactElement<ITableProps>> = [];

        React.Children.forEach(this.props.children, (child) => {

            if (React.isValidElement<ITableProps>(child)){

                if (child.type === THead && child.props.children){

                    const row = this.getRowChildren(child.props.children);
                    headStrings = this.collectData(row);

                    const headers = this.getRowChildren(row);

                    const callbackHeaders = headers.map((header: React.ReactElement<IHeaderProps>, i: number) => {
                        return (
                            <TH
                                key = {i}
                                className = {header.props.className}
                                style = {header.props.style}
                                onClick = {this.onClickHeader(header)}
                            >
                                {header.props.children}
                            </TH>
                        );
                    });

                    newHeaders = (
                        <THead
                            className = {child.props.className}
                            style = {child.props.style}
                        >
                            <TR>
                                {callbackHeaders}
                            </TR>
                        </THead>
                    );

                    // newHeaders = child;
                }

                if (child.type === TBody){

                    const data = this.getRowChildren(child.props.children);
                    data.map((row: any, i: number) => {
                        const cells = this.getRowChildren(row);
                        cells.map((cell: any, index: number) => {
                            newCells.push(
                                <TD
                                    key = {index}
                                    className = {cell.props.className}
                                    style = {cell.props.style}
                                >
                                    <div className = {styles.tdContent}>
                                        <span className={styles.heading}>{headStrings[index]}</span>
                                        <span className={styles.content}>{cell.props.children}</span>
                                    </div>
                                </TD>
                            );
                        });
                        newRows.push(
                            <TR
                                key = {i}
                                className = {row.props.className}
                                style = {row.props.style}
                            >
                                {newCells}
                            </TR>
                        );
                        newCells = [];
                    });

                    newBody.push(
                        <TBody
                            key = {0}
                            className = {child.props.className}
                            style = {child.props.style}
                        >
                            {newRows}
                        </TBody>
                    );

                }

            }

        });

        return {newHeaders, newBody};

    }

    private getRowChildren = (row: any) => {
        if (Array.isArray(row)){
            return row;
        } else {
            return row.props.children;
        }
    }

    private collectData(headersArray: Array<React.ReactElement<ITableProps>>){

        const headers: string[] = [];

        headersArray.map((header) => {
            headers.push(header.props.children);
        });

        return headers;
    }

    private onClickHeader = (header: React.ReactElement<IHeaderProps>) => {
        return (e: React.MouseEvent<HTMLElement>) => {
            header.props.onClick && header.props.onClick(e);
        };
    }

}

export const THead: React.StatelessComponent<ITableProps> = (props) => {

    return(
        <thead className={`THead-component ${styles.thead}`} style={props.style}>
            {props.children}
        </thead>
    );
};

export const TH: React.StatelessComponent<IHeaderProps> = (props) => {

    return(
        <th className={`TH-component ${styles.th}`} style={props.style} onClick={props.onClick}>
            {props.children}
        </th>
    );
};

export const TBody: React.StatelessComponent<ITableProps> = (props) => {

    return(
        <tbody className={`TBody-component ${styles.tbody}`} style={props.style}>
            {props.children}
        </tbody>
    );
};

export const TR: React.StatelessComponent<ITableProps> = (props) => {

    return(
        <tr className={`TR-component ${styles.tr}`} style={props.style}>
            {props.children}
        </tr>
    );
};

export const TD: React.StatelessComponent<ITableProps> = (props) => {

    return(
        <td className={`TD-component ${styles.td}`} style={props.style}>
            {props.children}
        </td>
    );
};
