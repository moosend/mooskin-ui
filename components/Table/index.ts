import Table, {TBody, TD, TH, THead, TR} from './Table';

Table.TBody = TBody;
Table.THead = THead;
Table.TR = TR;
Table.TD = TD;
Table.TH = TH;

export default Table;
export {TBody, TD, TH, THead, TR};
