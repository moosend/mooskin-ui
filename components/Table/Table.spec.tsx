import * as React from 'react';
import Table, {TBody, TD, TH, THead, TR} from './Table';

import { shallow } from 'enzyme';

describe('Table', () => {

    test('renders correctly', () => {

        const tree = shallow(
            <Table>
                <THead>
                    <TR>
                        <TH>th 1</TH>
                        <TH>th 2</TH>
                        <TH>th 3</TH>
                    </TR>
                </THead>
                <TBody>
                    <TR>
                        <TD>td 1</TD>
                        <TD>td 2</TD>
                        <TD>td 3</TD>
                    </TR>
                    <TR>
                        <TD>td 1</TD>
                        <TD>td 2</TD>
                        <TD>td 3</TD>
                    </TR>
                </TBody>
            </Table>
        );
        expect(tree).toMatchSnapshot();
    });

    test('callback when header is clicked', () => {

        const func = jest.fn();

        const tree = shallow(
            <Table>
                <THead>
                    <TR>
                        <TH onClick = {func}>th 1</TH>
                        <TH>th 2</TH>
                        <TH>th 3</TH>
                    </TR>
                </THead>
                <TBody>
                    <TR>
                        <TD>td 1</TD>
                        <TD>td 2</TD>
                        <TD>td 3</TD>
                    </TR>
                    <TR>
                        <TD>td 1</TD>
                        <TD>td 2</TD>
                        <TD>td 3</TD>
                    </TR>
                </TBody>
            </Table>
        );

        tree.find(TH).first().simulate('click');
        expect(func).toHaveBeenCalled();

    });

});
