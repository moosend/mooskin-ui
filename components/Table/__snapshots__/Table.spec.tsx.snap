// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders correctly 1`] = `
<div>
  <table
    className="table "
    style={Object {}}
  >
    <THead>
      <TR>
        <TH
          onClick={[Function]}
        >
          th 1
        </TH>
        <TH
          onClick={[Function]}
        >
          th 2
        </TH>
        <TH
          onClick={[Function]}
        >
          th 3
        </TH>
      </TR>
    </THead>
    <TBody>
      <TR>
        <TD>
          <div
            className="tdContent"
          >
            <span
              className="heading"
            >
              th 1
            </span>
            <span
              className="content"
            >
              td 1
            </span>
          </div>
        </TD>
        <TD>
          <div
            className="tdContent"
          >
            <span
              className="heading"
            >
              th 2
            </span>
            <span
              className="content"
            >
              td 2
            </span>
          </div>
        </TD>
        <TD>
          <div
            className="tdContent"
          >
            <span
              className="heading"
            >
              th 3
            </span>
            <span
              className="content"
            >
              td 3
            </span>
          </div>
        </TD>
      </TR>
      <TR>
        <TD>
          <div
            className="tdContent"
          >
            <span
              className="heading"
            >
              th 1
            </span>
            <span
              className="content"
            >
              td 1
            </span>
          </div>
        </TD>
        <TD>
          <div
            className="tdContent"
          >
            <span
              className="heading"
            >
              th 2
            </span>
            <span
              className="content"
            >
              td 2
            </span>
          </div>
        </TD>
        <TD>
          <div
            className="tdContent"
          >
            <span
              className="heading"
            >
              th 3
            </span>
            <span
              className="content"
            >
              td 3
            </span>
          </div>
        </TD>
      </TR>
    </TBody>
  </table>
</div>
`;
